### ISSUE TITLE
Question about determinant in KLD loss

### ISSUE THREAD

**Issue by nmerrill67** [jramapuram/CVAE#1]
Hello, the formula for the KLD includes the log of the determinant of the covariance, however I see that you have it as a sum. I attempted to change it to a product, since we just have the matrix diagonal, but it give nan or inf values. I also noticed that other VAE implementations do the same. Could you perhaps elaborate on why this works if you have any insight?

**Comment by sujithvemi** [comment]
This question is very old, but just posting an answer for anyone who has this doubt in future.

The KL divergence actually has a logarithm of determinant. You are right about the determinant being the product because it is diagonal but the logarithm of that product just gives you the sum of logarithms of variances.

**Comment by sujithvemi** [comment]
> Hello, the formula for the KLD includes the log of the determinant of the covariance, however I see that you have it as a sum. I attempted to change it to a product, since we just have the matrix diagonal, but it give nan or inf values. I also noticed that other VAE implementations do the same. Could you perhaps elaborate on why this works if you have any insight?

 This question is very old, but just posting an answer for anyone who has this doubt in future.

The KL divergence actually has a logarithm of determinant. You are right about the determinant being the product because it is diagonal but the logarithm of that product just gives you the sum of logarithms of variances.

**Comment by jramapuram** [comment]
FYI: This codebase is pretty old. I recommend using tensorflow distributions or pytorch distributions instead of this solution, but @sujithvemi is correct: log(det(isotropic_gauss)) = sum of the diags.

I am currently using pytorch actively and you can spinup a VAE, CVAE, VRNN with this: https://github.com/jramapuram/ml_base
